### PR TITLE
Faster module for resampling time series.

### PR DESCRIPTION
**Pull request by small-yellow-duck**
The proposed Resampler routine is more efficient than the existing Resample
module for resampling time series signals. Speed improvements are obtained by splitting
the signal into blocks where there are 'input_sr' input samples and 'output_sr' output
samples. Each block is treated with a convolution mapping 'input_sr' input channels to
'output_sr' output channels per time step. The existing Resample module uses a for loop
to iterate over the first index where each filter is applied, but this implementation is
fully convolutional.

The module is based on

https://github.com/danpovey/filtering/blob/master/lilfilter/resampler.py

with improvements to include additional filter types and input parameters that align
with the librosa api.

https://github.com/pytorch/audio/issues/908

- [x] inputs where the sequence length is not a whole number multiple of `input_sr` should be padded with zeros
- [x] adapt module to deal with multi-channel input
- [ ] adapt to deal with non-integer sample rates?

Should tests try to reproduce signals resampled with [scipy.signal.resample](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.resample.html) ?

### CODE REVIEW

File: torchaudio/transforms.py [hunk 1]
```diff
@@ -9,6 +9,9 @@
 from torchaudio import functional as F
 from torchaudio.compliance import kaldi
 
+import numpy as np
+from scipy import special
```
Comment (mthrok) [inline torchaudio/transforms.py:13]: Hi @small-yellow-duck 

Thanks for opening a PR. I will try this out to think how to test. I have not looked into the detail yet but `numpy` and `scipy` are not a mandatory dependencies of `torchaudio`.

We have a mechanism to differ loading the optional dependencies but it would be nice if we can implement without them. Quick googling tells that `torch` has `heaviside` function, so removing `numpy` appears plausible. We need to think what to do for `scipy` dependency.

Comment (small-yellow-duck) [reply torchaudio/transforms.py:13]: `scipy.special.i0` is used to compute the kaiser window for `kaiser`, `kaiser_best` and `kaiser_fast`. If scipy dependency is not desirable, then one option would be to just allow `hann` window mode. (the original goal of including `kaiser` windows was to try to duplicate the behaviour of librosa, but I don't think the results of librosa and this proposed routine are identical to numeric precision, so there is likely some other detail about how the librosa filters were generated that I haven't managed to duplicate). If the routine only needs to support `hann` windows, then I agree that it should be possible to remove numpy and scipy as dependencies.

I also see that there is an issue suggesting that pytorch add the modified bessel functions, although there doesn't seem to be a PR to merge the code into the pytorch codebase: https://github.com/pytorch/pytorch/issues/7815 . If the code to calculate the equivalent of `scipy.special.i0` is added to pytorch, then it should also be possible to include `kaiser` windows.

Comment (small-yellow-duck) [reply torchaudio/transforms.py:13]: I realized that sinc can be written in terms of the gamma function - and pytorch has a routine to compute the log of the gamma function (`torch.lgamma`)! So I think it should be possible to excise numpy as a dependency as well.... (see output 10 in https://mathworld.wolfram.com/SincFunction.html)

**Comment by small-yellow-duck** [comment]
Any comments to the following changes? @mthrok 

- Removing numpy and scipy as dependencies means that the resampler will no longer support `kaiser` windows. `kaiser` windows can be supported if there is a pytorch definition for `scipy.special.i0`.

- In order to remove the numpy dependency, the sinc function is now defined as torch.exp(-torch.lgamma(1-x) - torch.lgamma(1+x)). This is not a good solution because the `exp` introduces numerical errors. A solution is to have a pytorch definition based on a taylor series expansion of `sinc`.

- The `general` mode in forward no longer crops out the portion of the signal that is longer than `(seq_len//input_sr) * input_sr`. (YAY)

I am still thinking about how to support non-integer sample rates. 
Further modifications to support multi-channel signals is desirable.

File: torchaudio/transforms.py [hunk 2]
```diff
@@ -610,6 +610,282 @@ def forward(self, waveform: Tensor) -> Tensor:
...
+        using convolutional code, imagining there are `input_sr`
+        input channels and `output_sr` output channels per time step.
+
+        """
+        assert isinstance(input_sr, int) and isinstance(output_sr, int)
+        if input_sr == output_sr:
+            self.resample_type = 'trivial'
+            return
+
+        def gcd(a, b):
```
Comment (mthrok) [inline torchaudio/transforms.py:669]: Kind of surprise to me but `torch` has gcd. https://pytorch.org/docs/master/generated/torch.gcd.html#torch-gcd

File: torchaudio/transforms.py [hunk 3]
```diff
@@ -610,6 +610,282 @@ def forward(self, waveform: Tensor) -> Tensor:
...
+
+
+        def kaiser_window(a, beta):
+            #TODO: the kaiser window can be added once scipy.special.i0 has been implement in torch
+            #w = scipy.special.i0(beta * np.sqrt(np.clip(1 - ((a - 0.0) / 1.0) ** 2.0, 0.0, 1.0))) / special.i0(beta)
+            #return np.heaviside(1 - np.abs(a), 0.0) * w
+            return None
+
+
+        def sinc(x):
```
Comment (mthrok) [inline torchaudio/transforms.py:760]: Looks like `sinc` function is under development.

**Review by mthrok — COMMENTED** [review]
Hi @small-yellow-duck 

Sorry for the late reply. I took a look at your PR and it was surprisingly straightforward to follow. I have not got to the point of thinking of testing phase, but I made couple of comments.

Specifically, `torch` seems to already have `kaiser_window` and `hann_window`. Did you look at them? `sinc` is also under development https://github.com/pytorch/pytorch/pull/44713 so it might become available rather soon.

On the interface design aspect, since this module has four distinctive operation modes (`trivial`, `integer_downsample`, `integer_upsample` and `general`), it might be a better to split the resampling class into 4 different and introduce a factory function. cc @cpuhrsch

File: torchaudio/transforms.py [hunk 4]
```diff
@@ -610,6 +610,282 @@ def forward(self, waveform: Tensor) -> Tensor:
...
+        """
+        Resample the data
+
+        Args:
+         input: a torch.Tensor with the same dtype as was passed to the
+           constructor.
+         There must be 2 axes, interpreted as (minibatch_size, sequence_length)...
+         the minibatch_size may in practice be the number of channels.
+
+         TODO: make default input dim (minibatch_size, channels, seq_len)?
```
[the remark this answers is not in the document]
Comment (small-yellow-duck) [reply torchaudio/transforms.py:823]: I also like the flexibility of having 2d or 3d inputs.

File: torchaudio/transforms.py [hunk 5]
```diff
@@ -610,6 +610,282 @@ def forward(self, waveform: Tensor) -> Tensor:
...
+
+            # num_blocks = seq_len // self.input_sr
+            # if num_blocks == 0:
+            #     #if signal is shorter than the filter, then pad signal with zeros.
+            #     zeros = torch.zeros(data.size()[0:-1]+(self.input_sr - seq_len,), dtype=data.dtype)
+            #     data = torch.cat([data, zeros], dim=-1)
+            #     num_blocks = 1
+
+            # data = data[:, 0:(num_blocks*self.input_sr)]  # Truncate input
+            data = data[:, 0:(num_blocks * self.input_sr)].view(minibatch_size, num_blocks, self.input_sr)
```
[the remark this answers is not in the document]
Comment (small-yellow-duck) [reply torchaudio/transforms.py:873]: Nope! Let me tidy up this piece of code a bit more...

File: torchaudio/transforms.py (hunk 2 above)
Comment (small-yellow-duck) [reply torchaudio/transforms.py:669]: `gcd` is also in the standard python library. I'm fine to switch to the pytorch `gcd` but is there some underlying reason why it would compelling to do so?

File: torchaudio/transforms.py (hunk 3 above)
Comment (small-yellow-duck) [reply torchaudio/transforms.py:760]: Oh good. I'll leave the definition of the sinc function as it is for now, but the new function should be used when it's available.

File: torchaudio/transforms.py [hunk 6]
```diff
@@ -610,6 +610,282 @@ def forward(self, waveform: Tensor) -> Tensor:
...
+        # sign that the kernel_width term needs to be will depend on whether it's
+        # convolution or correlation, and the logic is tricky.. I will just find
+        # which sign works.
+
+        times = (
+            torch.arange(output_sr, dtype=dtype).reshape((output_sr, 1, 1)) / output_sr -
+            torch.arange(input_sr, dtype=dtype).reshape((1, input_sr, 1)) / input_sr -
+            (torch.arange(kernel_width, dtype=dtype).reshape((1, 1, kernel_width)) - blocks_per_side))
+
+        def hann_window(a):
```
[the remark this answers is not in the document]
Comment (small-yellow-duck) [reply torchaudio/transforms.py:740]: `torch.hann_window` doesn't have quite the same functionality as the `hann_window` function in the resample routine: the pytorch `hann_window` computes the filter for a fixed window size (the only input is the window size) without the ability to scale the period of the filter (ie to change the argument of `cos(2*pi*n/N)` to `cos(2*pi*n*a/N)`. Compare this to the behaviour of the `hann_window` function in the resampler, which takes the 3d `times` array and computes the value of the filter for each element.

File: torchaudio/transforms.py [hunk 7]
```diff
@@ -610,6 +610,282 @@ def forward(self, waveform: Tensor) -> Tensor:
...
+            This is applied elementwise to a, which should be a NumPy array.
+
+            The heaviside function returns (a > 0 ? 1 : 0).
+            """
+            #return np.heaviside(1 - np.abs(a), 0.0) * (0.5 + 0.5 * np.cos(a * np.pi))
+            pi = torch.acos(torch.zeros(1)).item() * 2
+            return torch.heaviside(1 - torch.abs(a), 0.0) * (0.5 + 0.5 * torch.cos(a * pi))
+
+
+        def kaiser_window(a, beta):
```
[the remark this answers is not in the document]
Comment (small-yellow-duck) [reply torchaudio/transforms.py:753]: same issue as with `hann_window` above... but the good news is that pytorch now has the `i0` function, so I can rewrite the `resampler` `kaiser_window` routine with that. https://pytorch.org/docs/master/generated/torch.i0.html#torch.i0

**Comment by small-yellow-duck** [comment]
Your idea is to have a base class `resample` that does the weight generation and which `resample_trivial`, `resample_integer_upsample`, `resample_integer_downsample` and `resample_general` can inherit from? I am pretty agnostic on whether to split the resampler into `resample_trivial`, `resample_integer_upsample`, `resample_integer_downsample` and `resample_general`: I'm happy to do it if people think it makes the code easier to understand. 

What are your thoughts on what the goal of tests should be? This issue has some discussion about trying to duplicate the behaviour of audio processing tools in scipy with pytorch.
https://github.com/pytorch/audio/issues/452

Another issue is that the resampler `hann_window` function needs to have a value for pi, but pytorch doesn't have a definition for pi. I've used a suggested workaround (`pi = 2.0*torch.acos(torch.zeros(1)).item()`, but there is also an open issue to make pytorch include pi: https://github.com/pytorch/pytorch/issues/19172

**Comment by small-yellow-duck** [comment]
The latest PR now has the ability to accept 2d or 3d inputs.

- Is there a compelling reason to switch from using `gcd` to using `torch.gcd`?
- Is there a better workaround for a definition of `pi`?
- Is it desirable to split the `resample` class into `resample_trivial`, `resample_integer_upsample`, `resample_integer_downsample` and `resample_general` and have each class inherit the weight initializations from `resample`?
- What should the goals of unit tests be? (with the existing definition of sinc and pi, we can't expect numeric agreement with other packages...) Perhaps one option is to downsample a signal and then upsample it again? (and vice versa?)
@mthrok 

I think there is also a minor issue with `general` mode padding the input with 0's so that the `seq_len` is an integer multiple of `input_sr`. When the kernel width is small, this is pretty safe to do, but if the kernel width is large, then the convolution will involve a lot of the padded zeros and the output of the convolution is not going to be scaled correctly (the output will be too small). It should be possible to see this by comparing a signal padded with zeros at the end (`resample(torch.cat(data, zeros), dim=-1)[:, 0:new_seq_len]`) to the same signal padded with zeros at the begining (`resample(torch.cat(zeros, data), dim=-1)[:, -new_seq_len:]`). Maybe there is a smart way to to construct an appropriately normalized kernel to take care of the last segment of the input `data[:, :, (seq_len//self.input_sr)*self.input_sr:]`, or the case where the `seq_len` is shorter than `self.input_sr`. (However, this kernel would have to be constructed during the `.forward` call as the size depends on the data `seq_len`.)

Finally, I think it would be good for somebody to really scrutinize the definition I have for `kaiser_window` to make sure that it is correct. :-)

**Comment by small-yellow-duck** [comment]
Any comments on the latest revision or the questions in the previous post? @mthrok

**Comment by mthrok** [comment]
Hi @small-yellow-duck 

Sorry for the late reply. I do not forget you and this PR, but I got caught up with some works related to the upcoming release. I will try my best to get back to you before the end of the week.

**Comment by mthrok** [comment]
Hi @small-yellow-duck

Sorry for the delayed response. I am still working on release-related task, please give me some more time before I can give feedback.

**Comment by small-yellow-duck** [comment]
No worries, thanks for keeping this in PR mind. :-)

**Comment by mthrok** [comment]
@small-yellow-duck 

Sorry for the late reply. torchaudio 0.7.0 is released and I was reviewing the project ideas. I posted an RFC #1000 regarding the faster I/O, which I have been working, then realized that in this work we can provide a resampling algorithm faster than the current Python implementation. Therefore I would like to hold this PR. I am so sorry for the efforts you have made based on the interaction here.

**Comment by small-yellow-duck** [comment]
I expect most people want to resample data as they load it and I agree that faster resampling with C++ on the cpu is the functionality most worth focusing on.

**Comment by faroit** [comment]
Also maybe its worth, comparing to https://github.com/adefossez/julius that was just released

**Comment by mthrok** [comment]
Closing the issue as it is mostly addressed by #1487. 
Please comment in #1487 if we are missing something.
Thank you for the contribution @small-yellow-duck
